precompiles/xtokens/src/lib.rs

```diff
@@ -293,4 +308,190 @@ where
+		ensure!(
+			max_assets >= non_mapped_currencies.len(),
+			gasometer.revert("More than max number of assets given")
+		);
+
+		let fee_item: u32 = input.read::<u32>(gasometer)?;
+
+		// read destination
+		let destination: MultiLocation = input.read::<MultiLocation>(gasometer)?;
+
+		let dest_weight: u64 = input.read::<u64>(gasometer)?;
+
+		let origin = Runtime::AddressMapping::into_account_id(context.caller);
+
+		// Build all currencies
+		let currencies: EvmResult<
+			Vec<(
+				<Runtime as orml_xtokens::Config>::CurrencyId,
+				XBalanceOf<Runtime>,
+			)>,
+		> = non_mapped_currencies
+			.iter()
+			.map(|currency| {
+				let address_as_h160: H160 = currency.address.clone().into();
+				let amount = currency.amount.clone().try_into().map_err(|_| {
+					gasometer.revert("Amount is too large for provided balance type")
+				})?;
+
+				Ok((
+					Runtime::account_to_currency_id(Runtime::AddressMapping::into_account_id(
+						address_as_h160,
+					))
+					.ok_or(gasometer.revert("cannot convert into currency id"))?,
+					amount,
+				))
+			})
+			.collect();
+
+		let currencies_non_result = currencies?;
+
+		let call = orml_xtokens::Call::<Runtime>::transfer_multicurrencies {
+			currencies: currencies_non_result,
+			fee_item,
+			dest: Box::new(VersionedMultiLocation::V1(destination)),
+			dest_weight,
+		};
+
+		RuntimeHelper::<Runtime>::try_dispatch(Some(origin).into(), call, gasometer)?;
+
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: Default::default(),
+			logs: Default::default(),
+		})
+	}
+
+	fn transfer_multi_assets(
+		input: &mut EvmDataReader,
+		gasometer: &mut Gasometer,
+		context: &Context,
+	) -> EvmResult<PrecompileOutput> {
+		input.expect_arguments(gasometer, 4)?;
+		let assets: Vec<EvmMultiAsset> = input.read::<Vec<EvmMultiAsset>>(gasometer)?;
+		let max_assets = MaxAssetsForTransfer::<Runtime>::get();
+
+		// We check this here so that we avoid iterating over the vec
+		// if the len is more than the max permitted
+		ensure!(
+			max_assets >= assets.len(),
+			gasometer.revert("More than max number of assets given")
+		);
+
+		let fee_item: u32 = input.read::<u32>(gasometer)?;
+
+		// read destination
+		let destination: MultiLocation = input.read::<MultiLocation>(gasometer)?;
+
+		let dest_weight: u64 = input.read::<u64>(gasometer)?;
+
+		let origin = Runtime::AddressMapping::into_account_id(context.caller);
+
+		let multiasset_vec: EvmResult<Vec<MultiAsset>> = assets
+			.iter()
+			.map(|evm_multiasset| {
+				let to_balance: u128 = evm_multiasset.amount.clone().try_into().map_err(|_| {
+					gasometer.revert("Amount is too large for provided balance type")
+				})?;
+				Ok((evm_multiasset.location.clone(), to_balance).into())
+			})
+			.collect();
+
+		// Since multiassets sorts them, we need to check whether the index is still correct,
+		// and error otherwise as there is not much we can do other than that
```

Comment:
IIRC, the sorting and deduping is a no-op outside of dev mode. If I'm right about that, we may want to manually ensure that it is sorted.

Reply:
Nevermind, this idea came from the variant of this fn named `from_sorted_and_deduplicated_skip_checks` which is obviously irrelevant here :)